third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -197,6 +197,17 @@ impl ControlFlowGraphSimplifierTransformation {
 /// 1. construct a `ControlFlowGraphCodeGenerator` from the code,
 /// 2. perform the transformation on the `ControlFlowGraphCodeGenerator`
 /// 3. generate the code back from the `ControlFlowGraphCodeGenerator`
+///
+/// Invariants:
+/// 1. Fields `code_blocks`, `successors`, and `predecessors` have the same set of keys.
+/// 2. Block `A` has sucessor block `B` iff block `B` has predecessor block `A`, according
+/// to fields `successors` and `predecessors` respectively.
+/// 3. In `code_blocks`, entry and exit blocks cannot have code (mapped to empty vectors in `code_blocks`),
+/// and all other blocks must have code.
+/// 4. Entry block has exactly one successor.
+/// 5. All blocks except for the exit block must have at least one successor.
```

Comment:
Should all blocks except the entry block have at least one predecessor?

Reply:
It's not required that each block (except the entry block) has a predecessor, for instance, there are unreachable (from entry) blocks.

Unreachable (from entry) blocks are also required to have successors. This is (implicitly) guaranteed by the `StacklessControlFlowGraph`, and maintained through our transformations.